### PR TITLE
feat: transcode in stats

### PR DESCRIPTION
**Pull request by guyfischman**
Adds transcoded videos storage usage to the server stats page, both for users and server total.

Mentioned in https://github.com/immich-app/immich/discussions/14093.

### CODE REVIEW

File: server/src/repositories/user.repository.ts [hunk 1]
```diff
@@ -117,6 +117,10 @@ export class UserRepository implements IUserRepository {
         'usageVideos',
       )
       .addSelect('users.quotaSizeInBytes', 'quotaSizeInBytes')
+      .addSelect(
+        `ARRAY_AGG(assets.encodedVideoPath) FILTER (WHERE assets.encodedVideoPath IS NOT NULL)`,
```
Comment (danieldietzler) [inline server/src/repositories/user.repository.ts:121]: I feel like this is gonna be veeeery slow

File: server/src/services/server.service.spec.ts [hunk 2]
```diff
@@ -177,8 +178,16 @@ describe(ServerService.name, () => {
     });
   });
 
+  vi.mock('fs', async () => {
+    const actual = (await vi.importActual<typeof fs>('fs'))!;
+    return {
+      ...actual,
+      statSync: vi.fn().mockReturnValue({ size: 500 }),
+    };
+  });
```
Comment (danieldietzler) [inline server/src/services/server.service.spec.ts:187]: This is now how you should be doing that. You should be mocking repo methods if you have to

File: server/src/services/server.service.ts [hunk 3]
```diff
@@ -131,11 +132,26 @@ export class ServerService extends BaseService {
       usage.usageVideos = user.usageVideos;
       usage.quotaSizeInBytes = user.quotaSizeInBytes;
 
+      for (const path of user.encodedVideoPaths) {
+        if (path.trim() === '') {
+          continue;
+        }
+
+        try {
+          const stats = fs.statSync(path);
```
Comment (danieldietzler) [inline server/src/services/server.service.ts:141]: Pretty sure we already have a repo method for that. If we don't, we should add one

Comment (mertalev) [inline server/src/services/server.service.ts:141]: This is terribly slow and will scale poorly as the number of videos increases, both in fetching every single path from the DB at once and in `stat`'ing them like this.

**Review by danieldietzler — APPROVED** [review]
I'm not even sure if we want this tbh. To me this sounds somewhat counterintuitive. When thinking about quota I think about how much data they are allowed to upload

**Review by danieldietzler — CHANGES_REQUESTED** [review]
Clicked the wrong button oof

**Review by mertalev — CHANGES_REQUESTED** [review]
I'm not fully convinced we need this. There is an elegance to only counting the original files - it aligns with how cloud solutions count usage, is independent from the particular settings of the admin and the immich release, and leads to a more intuitive quota for the user. This PR only changes the visible stats, but this is also confusing since there's a discrepancy between what is actually included in the quota and what is displayed.

Even if we come to an agreement there, the particular approach in this PR is too bolted-on and would need to be designed in a better way that doesn't require scanning the file system and encompasses other generated files like thumbnails.

**Comment by guyfischman** [comment]
> I'm not even sure if we want this tbh. To me this sounds somewhat counterintuitive. When thinking about quota I think about how much data they are allowed to upload
> I'm not fully convinced we need this. There is an elegance to only counting the original files - it aligns with how cloud solutions count usage, is independent from the particular settings of the admin and the immich release, and leads to a more intuitive quota for the user. This PR only changes the visible stats, but this is also confusing since there's a discrepancy between what is actually included in the quota and what is displayed.

I agree as regards the user's expectation. My change is only for the admin panel (at least that's how I intended it).

> I feel like this is gonna be veeeery slow
> This is terribly slow and will scale poorly as the number of videos increases, both in fetching every single path from the DB at once and in stat'ing them like this.

I agree, and now don't know why I did it this way to begin with. I think I didn't see the transcoded filesize in the db schema and wanted something that just worked now. I guess a more reasonable solution is to have the transcoding job update the asset table with the new file's size so it can be queried directly.

> Even if we come to an agreement there, the particular approach in this PR is too bolted-on and would need to be designed in a better way that doesn't require scanning the file system and encompasses other generated files like thumbnails.

Happy to include other things like thumbnails, database, and immich code itself tbh. I'd like the admin panel to just show everything immich is using. Transcoding just seemed like the biggest chunk.

**Comment by mertalev** [comment]
You're right - the file size for transcodes etc. isn't stored anywhere. I think it'd be reasonable to add a file size column in the `asset_files` table and move `encodedVideoPath` from `assets` over to this table. You could change the associated jobs to include file size when updating the DB (sharp's `toFile` provides this info and I believe fluent-ffmpeg does as well, so you shouldn't need separate `stat` calls). Getting this info would then just be a query on `asset_files` without touching the file system.

The only limitation of this is that it won't retroactively count files generated before this change, but I think this is okay as long as it's noted somewhere.

**Comment by alextran1502** [comment]
Hello, we have some incoming changes to the database that relate to files in the database. It will be cleaner to use that mechanism to report these stats. Hence, I am closing this PR for now. Sorry, and thank you for the contribution; for future reference, please contact us first on Discord if you plan on opening a larger PR.

**Comment by guyfischman** [comment]
> Hello, we have some incoming changes to the database that relate to files in the database. It will be cleaner to use that mechanism to report these stats. Hence, I am closing this PR for now. Sorry, and thank you for the contribution; for future reference, please contact us first on Discord if you plan on opening a larger PR.

Thanks - FTR I did contact on discord first :).

**Pull request closed without merging**